Pre-requisites to support basic Metal capture

Zorro666 opened the pull request:
## Description

Hooking and capture serialization for `MTLCommandBuffer` APIs `enqueue` and `waitUntilCompleted` (required to capture test programs with `enqueue` support). These chunks are added to the command buffer record in order that the chunks are only included in the capture when the command buffer is included. The replay support for these APIs treats them at the global scope and not replayed within the command buffer. The `enqueue` chunk might not be required or used by replay in the future (currently it is used to verify the order of command buffer reply matches what is expected).

Updated the `MetalCmdBufferRecordingInfo` struct to store the `CA::MetalLayer` & `WrappedMTLTexture` from the `MTL::Drawable` instead of the `MTL::Drawable`. This avoids validation asserts about accessing the texture of an `MTL::Drawable` after calling `MTLCommandBuffer::presentDrawable`. 

Changed member data `bool present` to `uint32_t flags` to be able to efficiently store more than one flag. Added `MetalCmdBufferStatus` enum to store different status flags.

Changed the return of `WrappedMTLCommandBuffer::GetCommandQueueReturn` to be a `WrappedMTLCommandQueue *` instead of `MTL::CommandQueue *`. It makes sense and is more natural to return the `m_CommandQueue` without casting it. The previous API usage was for the Objective C bridge layer which worked with `MTL::CommandQueue *. This is not required now the Objective C and C++ wrapping/unwrapping has been simplified.

This PR represents the final Objective C change in the current Metal development branch 🥳

Zorro666 committed:
MTLCommandBuffer API: enqueue, waitUntilCompleted

Hook and capture serialization for enqueue() & waitUntilCompleted()

Zorro666 committed:
Return the WrappedMTLCommandQueue from CmdBuffer

Previous API return goes back to usage for objective C calls and this is not required now the objective C and C++ wrapping/unwrapping has advanced and simplified.
More sense/natural/correct to return type without casting it.

Zorro666 committed:
Add MetalCmdBufferStatus enum

Zorro666 committed:
Update MetalCmdBufferRecordingInfo data

Switch to caching the CA::MetalLayer & WrappedMTLTexture from the MTL::Drawable.
This is to avoid validation asserts about accessing the texture of a MTL::Drawable after calling MTLCommandBuffer::presentDrawable.

Switch "bool present" to "uint32_t flags" to future proof being able to store multiple flags on the command buffer recoring info

baldurk commented on renderdoc/driver/metal/metal_resources.h:
```diff
@@ -124,23 +124,29 @@ inline MTL::Resource *Unwrap(WrappedMTLResource *obj)
   return Unwrap<MTL::Resource *>((WrappedMTLObject *)obj);
 }
 
+enum MetalCmdBufferStatus
```
Prefer using `enum class` for internal enums like this.

There are defines `BITMASK_OPERATORS` and `ITERABLE_OPERATORS` you can use for bitsets like this to ensure the enum class can still easily do bitwise operators.

Zorro666 committed:
Use capture mode directly : don't use local var

ie. replace

bool backframe = IsBackgroundCapturing(m_State);
if(backframe)

with

if(IsBackgroundCapturing(m_State))
